### PR TITLE
Heartbeat service telemetry collection

### LINKED ISSUE #30433: [Beats] Track metrics for Monitor Management Telemetry
Collect information from heartbeat to report back to the observability cluster for telemetry.

- Number of tests (count)
(a tile with a count showing the cumulative number of tests that have been executed on the service)
(a tile with a count showing the trailing 24h running average number of tests that have been executed on the service)
- Number of steps (for browser monitors) (count by number of steps)
(Histogram w/buckets on x axis of the count of steps and Y1 axis showing me number of monitors that match those numbers. (Preferably column)
- Weight of ES documents stored (count of docs, count of storage size)
Panel showing total count of docs ever generated and stored
Panel showing count of storage size used all time
_(Histogram w/buckets on x axis of the count of docs and Y1 axis showing me number of monitors that match those numbers.
- Running duration of tests (times)
Panel showing total duration of all tests run in total
_(Histogram w/buckets on x axis of the running time of each test run and Y1 axis showing me number of monitors that match those numbers. _

For the weight of ES documents stored, we discussed collecting the _raw_ data **per step** for browser monitors and per run for lightweight checks. If we don't, the size could be quite different/misleading between a 1 step monitor and a 20 step monitor (and we wouldn't have the data to know that). We can then use that data alongside our own internal tests to determine how much smaller on average that data will be once indexed into ES.

### PR DESCRIPTION
**Pull request by lucasfcosta**
> ⚠️ I'm moving this back to a draft state as I've recently spoken to @vigneshshanmugam and we'd like to keep changes as small as possible for the time being given: (1) we would like @andrewvc to review this once he's back and (2) we deemed these changes could be useful to improve Heartbeat's monitoring capabilities, especially considering how thorough they are.
> We decided to merge https://github.com/elastic/beats/pull/31405 just so that we can immediately have the metrics we need and then we'll look deeper into this PR.

## What does this PR do?

This PR adds the necessary metric collection mechanisms for https://github.com/elastic/beats/issues/30433.

## Why is it important?

It's necessary so that we have relevant stats in the telemetry cluster to analyse the service's behaviour.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Author's Checklist

- [ ] It's extremely important that we **test this PR E2E** as it deeply interferes with Heartbeat's capability of sending any stats to any ES instances.
- [ ] Please verify the output type of the data we're sending to our cluster to see if you agree with the way we're reporting data.
- [ ] Please ensure these are all the stats we need to perform the analysis we'll do for the service.
- [ ] As far as I understand, bytes written is already covered by https://github.com/elastic/beats/blob/main/libbeat/outputs/metrics.go#L41. I'd appreciate if someone could double-check I haven't missed anything here and advise on whether we need further changes to this.

### Open question

If we'll report all of the durations and steps the monitors ran, the monitoring output size will grow larger with time as we must register all of the duration buckets.

* Should we do something to prevent this from happening like more aggressive bucketing of possible durations and steps?
* Should we clear lists once they're reported once?

## How to test this PR locally

1. Run an ES instance, either locally or in the cloud
2. Create a `heartbeat.yml` configuration file and enable monitoring reporting
    ```yml
    monitoring:
      enabled: true
      elasticsearch:
        username: your_username
        password: your_password
    ```
3. Create a few monitors: both lightweight and heavyweight (both HTTP, and browser monitors, for example)
4. Build Heartbeat from this branch and start it
5. Query your cluster's `.monitoring-beats*` indices for this monitoring data and verify it contains the new fields for the step and durations of monitors.
  ```
  GET .monitoring-beats*/_search
  {
    "sort": { "timestamp": { "order":  "desc" }}
  }
  ```

## Related issues

Fixes #30433.

## Screenshots

Screenshot 2022-04-24 at 14 31 40

## Logs

A debug log indicating the fields have been sent:

```
{"log.level":"debug","@timestamp":"2022-04-23T18:23:49.848+0100","log.logger":"monitoring","log.origin":{"file.name":"processing/processors.go","file.line":209},"message":"Publish event: {\n  \"@timestamp\": \"2022-04-23T17:23:49.846Z\",\n  \"@metadata\": {\n    \"beat\": \"heartbeat\",\n    \"type\": \"_doc\",\n    \"version\": \"8.3.0\",\n    \"cluster_uuid\":
\"LI2Mb7skRcmzu4nuvKPiUQ\",\n    \"type\": \"beats_stats\",\n    \"interval_ms\": 10000,\n    \"params\": {\n      \"interval\": \"10s\"\n    }\n  },\n  \"metrics\": {\n    \"beat\": {\n      \"info\": {\n        \"uptime\": {\n          \"ms\": 40049\n        },\n        \"ephemeral_id\": \"121f5605-edaa-4d9c-b259-386f17b6e7ad\",\n        \"version\": \"8.3.0\"\
n      },\n      \"memstats\": {\n        \"memory_total\": 113862152,\n        \"memory_alloc\": 40542512,\n        \"memory_sys\": 70534152,\n        \"gc_next\": 43425440,\n        \"rss\": 88915968\n      },\n      \"cpu\": {\n        \"user\": {\n          \"ticks\": 983,\n          \"time\": {\n            \"ms\": 983\n          }\n        },\n        \"sys
tem\": {\n          \"ticks\": 186,\n          \"time\": {\n            \"ms\": 186\n          }\n        },\n        \"total\": {\n          \"time\": {\n            \"ms\": 1169\n          },\n          \"value\": 0,\n          \"ticks\": 1169\n        }\n      },\n      \"runtime\": {\n        \"goroutines\": 44\n      }\n    },\n    \"libbeat\": {\n      \"co
nfig\": {\n        \"scans\": 0,\n        \"reloads\": 0,\n        \"module\": {\n          \"stops\": 0,\n          \"running\": 0,\n          \"starts\": 0\n        }\n      },\n      \"output\": {\n        \"write\": {\n          \"errors\": 0,\n          \"bytes\": 2582611\n        },\n        \"read\": {\n          \"errors\": 0,\n          \"bytes\": 23497\
n        },\n        \"type\": \"elasticsearch\",\n        \"events\": {\n          \"duplicates\": 383,\n          \"active\": 0,\n          \"toomany\": 0,\n          \"batches\": 16,\n          \"total\": 715,\n          \"acked\": 332,\n          \"failed\": 0,\n          \"dropped\": 0\n        }\n      },\n      \"pipeline\": {\n        \"queue\": {\n
    \"max_events\": 4096,\n          \"acked\": 715\n        },\n        \"clients\": 4,\n        \"events\": {\n          \"active\": 0,\n          \"total\": 715,\n          \"filtered\": 0,\n          \"published\": 715,\n          \"failed\": 0,\n          \"dropped\": 0,\n          \"retry\": 3\n        }\n      }\n    },\n    \"heartbeat\": {\n      \"http\
": {\n        \"monitor_stops\": 0,\n        \"endpoint_starts\": 1,\n        \"endpoint_stops\": 0,\n        \"duration_histogram\": {\n          \"258\": 1\n        },\n        \"monitor_starts\": 1\n      },\n      \"icmp\": {\n        \"endpoint_starts\": 0,\n        \"endpoint_stops\": 0,\n        \"monitor_starts\": 0,\n        \"monitor_stops\": 0\n      }
,\n      \"tcp\": {\n        \"endpoint_stops\": 0,\n        \"monitor_starts\": 0,\n        \"monitor_stops\": 0,\n        \"endpoint_starts\": 0\n      },\n      \"browser\": {\n        \"monitor_starts\": 3,\n        \"monitor_stops\": 0,\n        \"endpoint_starts\": 3,\n        \"endpoint_stops\": 0,\n        \"duration_histogram\": {\n          \"10023\": 1
,\n          \"1164\": 1,\n          \"4768\": 1\n        },\n        \"steps_histogram\": {\n          \"1\": 1,\n          \"2\": 1,\n          \"3\": 1\n        }\n      },\n      \"scheduler\": {\n        \"tasks\": {\n          \"active\": 0,\n          \"waiting\": 0\n        },\n        \"jobs\": {\n          \"missed_deadline\": 0,\n          \"active\":
0\n        }\n      }\n    },\n    \"system\": {\n      \"load\": {\n        \"15\": 5.1177,\n        \"norm\": {\n          \"1\": 0.2659,\n          \"5\": 0.3255,\n          \"15\": 0.3199\n        },\n        \"1\": 4.2539,\n        \"5\": 5.2075\n      },\n      \"cpu\": {\n        \"cores\": 16\n      }\n    }\n  },\n  \"beat\": {\n    \"name\": \"Lucass-Ma
cBook-Pro.local\",\n    \"host\": \"Lucass-MacBook-Pro.local\",\n    \"uuid\": \"e585215b-0595-4ed2-9031-6f91a39aa294\",\n    \"type\": \"heartbeat\",\n    \"version\": \"8.3.0\"\n  }\n}","service.name":"heartbeat","ecs.version":"1.6.0"}
```

### CODE REVIEW

**Comment by lucasfcosta** [comment]
@elasticmachine /test

File: heartbeat/monitors/plugin/lightweight_duration_recorder.go [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+	durationHistogram *monitoring.UniqueList // histogram with the count of durations (in ms) for tests
+}
+
+func NewDurationRecorder(pluginName string, r *monitoring.Registry) DurationRegistryRecorder {
+	pluginRegistry := r.GetRegistry(pluginName)
+	if pluginRegistry == nil {
+		pluginRegistry = r.NewRegistry(pluginName)
+	}
+
+	durationHistogram := monitoring.NewUniqueList()
```
Comment (andrewvc) [inline heartbeat/monitors/plugin/lightweight_duration_recorder.go:40]: I'm wondering now if this is right, why should the list be unique for a histogram?

For a true histogram it's best to just use a fixed size array with the last bucket as an overflow for anomalous values. Not worth it at this point, but it lets you record data in a constant size data structure. For larger ranges of values you can map it to log values or similar.

**Review by andrewvc — COMMENTED** [review]
This mostly looks good, but TBH, it was very hard to follow with all the linting issues. Per your suggestion I think we should fix all the linting issues in one big new PR then rebase this on top of this to make this more tractable before I give a more thorough review.

The behavior seems good on first look. One trick for testing, by the way, is to use `output.file.path: /tmp/hbout`, then running heartbeat with `/heartbeat -e 2>&1 | jq .` to see the monitoring output

**Comment by andrewvc** [comment]
This now needs a rebase now that https://github.com/elastic/beats/pull/31503 is merged
